File: Data Structures & Algorithms/longest-common-prefix/submission-0.py
class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        prefix, curr = "", ""
        i, j = 0, 1

        while i < len(strs[0]):
            curr = strs[0][i]
            
            while j < len(strs) and i < len(strs[j]) and curr == strs[j][i]:
                j += 1

            if j != len(strs): 
                break

            prefix += curr
            j = 1
            i += 1

        return prefix

File: Data Structures & Algorithms/anagram-groups/submission-0.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        // map to collect all strings (in array form) and respective list collection of anagrams
        // key = String, value = list of 
        Map<String, List<String>> map = new HashMap<>();

        for (String str : strs) {
            // creating frequency array for key of map for each str in strs
            int[] freqArr = new int[26];
            // frequency array edited
            for (char ch : str.toCharArray()) {
                freqArr[ch - 'a']++;
            }

            // if string freqency array doesnt already exist, create an empty one
            String key = Arrays.toString(freqArr);
            map.putIfAbsent(key, new ArrayList<>());

            // add current str to existing list associated with freqArr
            map.get(key).add(str);
        }

        return new ArrayList<>(map.values());
    }
}
